src/sparse_blas/backends/cusparse/cusparse_handles.cpp

```diff
@@ -0,0 +1,520 @@
+            // Ensure that a cusparse handle is created before any other cuSPARSE function is called.
+            sc.get_handle(queue);
+            if (dvhandle->size != size) {
+                CUSPARSE_ERR_FUNC(cusparseDestroyDnVec, dvhandle->backend_handle);
+                auto cuda_value_type = CudaEnumType<fpType>::value;
+                CUSPARSE_ERR_FUNC(cusparseCreateDnVec, &dvhandle->backend_handle, size,
+                                  sc.get_mem(acc), cuda_value_type);
+                dvhandle->size = size;
+            }
+            else {
+                CUSPARSE_ERR_FUNC(cusparseDnVecSetValues, dvhandle->backend_handle,
+                                  sc.get_mem(acc));
+            }
+            dvhandle->set_buffer(val);
+        });
+    });
+    event.wait_and_throw();
+}
+
+template <typename fpType>
+void set_dense_vector_data(sycl::queue &queue, dense_vector_handle_t dvhandle, std::int64_t size,
+                           fpType *val) {
+    detail::check_can_reset_value_handle<fpType>(__func__, dvhandle, false);
+    auto event = queue.submit([&](sycl::handler &cgh) {
+        submit_host_task(cgh, queue, [=](CusparseScopedContextHandler &sc) {
+            // Ensure that a cusparse handle is created before any other cuSPARSE function is called.
+            sc.get_handle(queue);
+            if (dvhandle->size != size) {
+                CUSPARSE_ERR_FUNC(cusparseDestroyDnVec, dvhandle->backend_handle);
+                auto cuda_value_type = CudaEnumType<fpType>::value;
+                CUSPARSE_ERR_FUNC(cusparseCreateDnVec, &dvhandle->backend_handle, size,
+                                  sc.get_mem(val), cuda_value_type);
+                dvhandle->size = size;
+            }
+            else {
+                CUSPARSE_ERR_FUNC(cusparseDnVecSetValues, dvhandle->backend_handle,
+                                  sc.get_mem(val));
+            }
+            dvhandle->set_usm_ptr(val);
+        });
+    });
+    event.wait_and_throw();
+}
+
+FOR_EACH_FP_TYPE(INSTANTIATE_DENSE_VECTOR_FUNCS);
+
+sycl::event release_dense_vector(sycl::queue &queue, dense_vector_handle_t dvhandle,
+                                 const std::vector<sycl::event> &dependencies) {
+    // Use dispatch_submit_impl_fp to ensure the backend's handle is kept alive as long as the buffer is used
+    auto functor = [=](CusparseScopedContextHandler &) {
+        CUSPARSE_ERR_FUNC(cusparseDestroyDnVec, dvhandle->backend_handle);
+        delete dvhandle;
+    };
+    return dispatch_submit_impl_fp(__func__, queue, dependencies, functor, dvhandle);
+}
+
+// Dense matrix
+template <typename fpType>
+void init_dense_matrix(sycl::queue &queue, dense_matrix_handle_t *p_dmhandle, std::int64_t num_rows,
+                       std::int64_t num_cols, std::int64_t ld, layout dense_layout,
+                       sycl::buffer<fpType, 1> val) {
+    auto event = queue.submit([&](sycl::handler &cgh) {
```

Comment:
If matrix creation was supposed to be a `host`-side allocation operation, why do we need to do this inside a `host_task`? Is it because this is `sycl::buffer` and we need to wait on dependencies (if any) through the `get_access` call first?

Reply:
The buffer API needs this to handle the dependencies as you noted. Both USM and buffer API also need the host task to get a `sycl::interop_handle` object, needed to get the native context which is the key used to get the global cuSPARSE handle. According to the cuSPARSE documentation the global cuSPARSE handle must be created ["before any other cuSPARSE API function is invoked"](https://docs.nvidia.com/cuda/cusparse/index.html#cusparsecreate) even if that handle is not used.

Reply:
In case of BLAS and some other domains, there are no data structures to permanently store the `cublasHandle_t` or equivalents. In case of the sparse BLAS domain, however, all our APIs involve at least one data structure (sparse/dense matrix/vector handle). Is there a reason why we can't just have a `shared_ptr<cusparseHandle_t>` be a member of the matrix and vector handles, initialize/store it during calls to `init_matrix_handle`/equivalents, and then use it as needed instead of going through `host_task`->`interop_handle`->`cusparseHandle_t`?

Reply:
Also answering for the discussion [below](https://github.com/oneapi-src/oneMKL/pull/527#discussion_r1757501071).
The `optimize` functions and compute functions need to return a `sycl::event` so they will still need the host task. It's not clear how much difference it would make to use a global handle stored in the matrix handle like you suggest. This is a large change so I have some concerns that it will be too late to implement and review before 2025.0 is released.
I first tried to avoid the host_task in `set_*_data` function as they don't need the global handle but this is failing and will need to investigate more.

Reply:
@gajanan-choudhary from a first glance I don't see why remove the `host_task` is not working. Since this is only an optimization would you be ok if we create a separate issue to improve this after this is merged? I don't think this should block this PR and the rocSPARSE one. We are still aiming to merge them both before the end of October.

Reply:
Yeah, I figured it would be a large change to store it in the matrix handle. No need to create the `host_task`-not-working situation a separate issue.